### PR TITLE
[WIP][JIT] Add benchmarking support of NV Fuser with FP16 dtype support

### PR DESCRIPTION
**Pull request by kevinstephano**
Modified files in `benchmarks/tensorexpr` to add support for NVIDIA's Fuser for the jit compiler. 

This support has some modifications besides adding an option to support the NVIDIA fuser:

* Adds FP16 Datatype support
* Fixes SOL/Algo calculations to generally use the data type instead of being fixed to 4 bytes
* Adds IR printing and kernel printing knobs
* Adds a knob `input_iter` to create ranges of inputs currently only for reductions
* Adds further reduction support for Inner and Outer dimension reductions that are compatible with the `input_iter` knob.
* Added `simple_element`, `reduce2d_inner`, and `reduce2d_outer` to isolate performance on elementwise  and reduction operations in the most minimal fashion.

### CODE REVIEW

**Comment by csarofeen** [comment]
We still have some work to do cleaning this up, but wanted to test the waters for additions here with @ZolotukhinM and @bertmaher

File: benchmarks/tensorexpr/__main__.py [hunk 1]
```diff
@@ -133,13 +163,59 @@ def set_global_threads(num_threads):
...
+                    )
+            bench.run(args)
+
+    def run_with_input_iter(bench_cls, input_iter, allow_skip=True):
+        tensor_dim_specs = input_iter.split(',')
+        tensor_dim_specs = [dim.split(':') for dim in tensor_dim_specs]
+
+        configs = []
+        for dim in tensor_dim_specs :
+            assert len(dim) == 3, "There should be three dimensions for: start, stop, inc!"
```
Comment (bertmaher) [inline benchmarks/tensorexpr/__main__.py:196]: minor, this block of code would be clearer if `dim` were unpacked as `start, stop, inc = dim`

File: benchmarks/tensorexpr/benchmark.py [hunk 2]
```diff
@@ -82,6 +85,20 @@ def compute_workload(self):
...
         return None
 
+    @staticmethod
+    def input_iterable():
+        """A benchmark child class should return true if it utilizes the input iter arg"""
+        return False
+
+    def dtype_to_bytes(self) :
+        import torch
+        if self.dtype == torch.float32 :
```
Comment (bertmaher) [inline benchmarks/tensorexpr/benchmark.py:95]: You could ask torch for the size of a dtype by doing `torch.tensor(0, dtype=self.dtype).element_size()`.  I have mixed feelings about creating a tensor just for this purpose, but it does at least keep the source of truth in `torch` instead of our benchmark harness :-p.

**Review by bertmaher — APPROVED** [review]
Cool!  I have just a couple suggestions which you're welcome to take or leave :).  (And, sorry this benchmark harness is maybe a bit clunky.  Hopefully this change wasn't too onerous to write!)

File: benchmarks/tensorexpr/__main__.py (hunk 1 above)
Comment (kevinstephano) [reply benchmarks/tensorexpr/__main__.py:196]: I like your way better.  Thanks for the suggestion!

File: benchmarks/tensorexpr/benchmark.py (hunk 2 above)
Comment (kevinstephano) [reply benchmarks/tensorexpr/benchmark.py:95]: I am perfectly fine with this.  I just added the method as I wasn't finding an ideal choice.  One other note is that I think the Dtype based memory calculations will need to change again.  I was about to add some ops that had mixed precision and quickly realized that generically multiplying by a type size was not going to work.

**Comment by kevinstephano** [comment]
@bertmaher I was curious why you chose to using numpy for reference checking versus just Pytorch eager mode?

**Comment by bertmaher** [comment]
@kevinstephano I think comparing with eager mode is definitely the right way to go!  @zheng-xq and I started this benchmark harness in a different repo where we were comparing several different frameworks (tf, tvm, etc.) in addition to pytorch and numpy was a sort of least common denominator to use as a reference.  But that attempt fell by the wayside and we now use it for pytorch exclusively, so we should just compare with eager :-).

**Comment by kevinstephano** [comment]
Hey @bertmaher, I think this PR is ready to be merged unless you have any more concerns/issues.
